flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -954,6 +954,27 @@ public void run() {
 							owner.getName(), checkpointMetaData.getCheckpointId(), asyncDurationMillis);
 				}
 			} catch (Exception e) {
+				// clean up ongoing operator snapshot results and non partitioned state handles
+				for (OperatorSnapshotResult operatorSnapshotResult : snapshotInProgressList) {
+					if (operatorSnapshotResult != null) {
+						try {
+							operatorSnapshotResult.cancel();
+						} catch (Exception cancelException) {
+							e.addSuppressed(cancelException);
+						}
+					}
+				}
+
+				for (StreamStateHandle nonPartitionedStateHandle : nonPartitionedStateHandles) {
```

Comment:
Having a second look, I think the other state objects don't have to be cleaned up, because they are still owned by the corresponding `OperatorSnapshotResult` which is cleaned up in case of a failure or `close` operation.